src/axom/core/FlatMap.hpp

```diff
@@ -0,0 +1,818 @@
+  friend class IteratorImpl;
+
+  using MixedHash = detail::flat_map::HashMixer64<KeyType, Hash>;
+
+public:
+  using key_type = KeyType;
+  using mapped_type = ValueType;
+  using size_type = IndexType;
+  using value_type = KeyValuePair;
+  using iterator = IteratorImpl<false>;
+  using const_iterator = IteratorImpl<true>;
+
+  /*!
+   * \brief Constructs a FlatMap with no elements.
+   */
+  FlatMap() : FlatMap(MIN_NUM_BUCKETS) { }
+
+  /*!
+   * \brief Constructs a FlatMap with at least a given number of buckets.
+   *
+   * \param [in] bucket_count the minimum number of buckets to allocate
+   */
+  explicit FlatMap(IndexType bucket_count);
+
+  /*!
+   * \brief Constructs a FlatMap with a range of elements.
+   *
+   * \param [in] first iterator pointing to the beginning of the range
+   * \param [in] last iterator pointing to the end of the range
+   * \param [in] bucket_count minimum number of buckets to allocate (optional)
+   */
+  template <typename InputIt>
+  FlatMap(InputIt first, InputIt last, IndexType bucket_count = -1)
+    : FlatMap(std::distance(first, last), first, last, bucket_count)
+  { }
+
+  /*!
+   * \brief Constructs a FlatMap with a range of elements.
+   *
+   * \param [in] init a list of pairs to initialize the map with
+   * \param [in] bucket_count minimum number of buckets to allocate (optional)
+   */
+  explicit FlatMap(std::initializer_list<value_type> init,
+                   IndexType bucket_count = -1)
+    : FlatMap(init.begin(), init.end(), bucket_count)
+  { }
+
+  /*!
+   * \brief Move constructor for a FlatMap instance.
+   *
+   * \param other the FlatMap to move data from
+   */
+  FlatMap(FlatMap&& other) : FlatMap() { swap(other); }
+
+  /*!
+   * \brief Move assignment operator for a FlatMap instance.
+   *
+   * \param other the FlatMap to move data from
+   */
+  FlatMap& operator=(FlatMap&& other) { swap(other); }
+
+  /*!
+   * \brief Copy constructor for a FlatMap instance.
+   *
+   * \param other the FlatMap to copy data from
+   *
+   * \pre KeyType must be copy-constructible
+   * \pre ValueType must be copy-constructible
+   */
+  FlatMap(const FlatMap& other)
+    : m_numGroups2(other.m_numGroups2)
+    , m_size(other.m_size)
+    , m_metadata(other.m_metadata)
+    , m_buckets(other.m_buckets.size())
+    , m_loadCount(other.m_loadCount)
+  {
+    static_assert(std::is_copy_constructible<KeyType>::value,
+                  "Cannot copy an axom::FlatMap when key type is not "
+                  "copy-constructible.");
+    static_assert(std::is_copy_constructible<ValueType>::value,
+                  "Cannot copy an axom::FlatMap when value type is not "
+                  "copy-constructible.");
+    // Copy all elements.
+    IndexType index = this->nextValidIndex(m_metadata, NO_MATCH);
+    while(index < bucket_count())
+    {
+      new(&m_buckets[index].data) KeyValuePair(other.m_buckets[index].get());
+      index = this->nextValidIndex(m_metadata, index);
```

Comment:
Should we be concerned about repeatedly constructing `ArrayView` for `m_metadata` in this inner loop?